rtnetlink/src/route/add.rs

```diff
@@ -0,0 +1,153 @@
+use futures::stream::StreamExt;
+use std::net::IpAddr;
+
+use netlink_packet_route::{
+    nlas::route::Nla, NetlinkMessage, NetlinkPayload, RouteKind, RouteMessage, RouteProtocol,
+    RouteScope, RouteTable, RtnlMessage, AF_INET, AF_INET6, NLM_F_ACK, NLM_F_CREATE, NLM_F_EXCL,
+    NLM_F_REQUEST,
+};
+
+use crate::{Error, ErrorKind, Handle};
+
+fn addr_octets(addr: &IpAddr) -> Vec<u8> {
+    match addr {
+        IpAddr::V4(addr) => addr.octets().to_vec(),
+        IpAddr::V6(addr) => addr.octets().to_vec(),
+    }
+}
+
+/// A request to create a new route. This is equivalent to the `ip route add` commands.
+pub struct RouteAddRequest {
+    handle: Handle,
+    message: RouteMessage,
+}
+
+impl RouteAddRequest {
+    pub(crate) fn new(handle: Handle) -> Self {
+        let mut message = RouteMessage::default();
+
+        message.header.table = RouteTable::Main;
+        message.header.protocol = RouteProtocol::Static;
+        message.header.scope = RouteScope::Universe;
+        message.header.kind = RouteKind::Unicast;
+
+        RouteAddRequest { handle, message }
+    }
+
+    fn set_addr_family(&mut self, addr: &IpAddr) {
+        match addr {
+            IpAddr::V4(_)
+                if self.message.header.address_family == 0
+                    || self.message.header.address_family == AF_INET as u8 =>
+            {
+                self.message.header.address_family = AF_INET as u8;
+            }
+            IpAddr::V6(_)
+                if self.message.header.address_family == 0
+                    || self.message.header.address_family == AF_INET6 as u8 =>
+            {
+                self.message.header.address_family = AF_INET6 as u8;
+            }
+            _ => panic!("All addresses within a routing request must be either IPv4 or IPv6."),
+        };
+    }
+
+    /// Sets the input interface index.
+    pub fn input_interface(mut self, index: u32) -> Self {
+        self.message.nlas.push(Nla::Iif(index));
+        self
+    }
+
+    /// Sets the output interface index.
+    pub fn output_interface(mut self, index: u32) -> Self {
+        self.message.nlas.push(Nla::Oif(index));
+        self
+    }
+
+    /// Sets the source address prefix.
+    ///
+    /// # Panics
```

Comment:
I don't think we should panic here. We should warn the user that mixing IPv4 and IPv6 addresses is an error, but I see no reason to panic. If the user does that, they will most likely receive a netlink error message anyway.

Reply:
Like above, I would prefer `source_prefix` and `destination_prefix` or `dest_prefix`. But no big deal if you feel strongly about the names.